src/test/java/com/aws/greengrass/security/SecurityServiceTest.java

```diff
@@ -0,0 +1,99 @@
+import com.aws.greengrass.security.exceptions.ServiceUnavailableException;
+import com.aws.greengrass.testcommons.testutilities.GGExtension;
+import org.hamcrest.collection.IsMapContaining;
+import org.hamcrest.collection.IsMapWithSize;
+import org.hamcrest.core.Is;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+import javax.net.ssl.KeyManager;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+@ExtendWith({MockitoExtension.class, GGExtension.class})
+class SecurityServiceTest {
+
+    private final SecurityService service = new SecurityService();
+
+    @Mock
+    private CryptoKeySpi mockKeyProvider;
+
+    @Test
+    void GIVEN_key_service_provider_WHEN_register_provider_THEN_succeed() throws Exception {
+        when(mockKeyProvider.supportedKeyType()).thenReturn("FILE");
+        service.registerCryptoKeyProvider(mockKeyProvider);
+        assertThat(service.getCryptoKeyProviderMap(), IsMapWithSize.aMapWithSize(1));
+        assertThat(service.getCryptoKeyProviderMap(), IsMapContaining.hasEntry(new CaseInsensitiveString("file"),
+                mockKeyProvider));
+    }
+
+    @Test
+    void GIVEN_key_service_provider_WHEN_register_provider_multiple_time_THEN_api_idempotent() throws Exception {
+        when(mockKeyProvider.supportedKeyType()).thenReturn("FILE");
+        service.registerCryptoKeyProvider(mockKeyProvider);
+        assertThat(service.getCryptoKeyProviderMap(), IsMapWithSize.aMapWithSize(1));
+        service.registerCryptoKeyProvider(mockKeyProvider);
+        assertThat(service.getCryptoKeyProviderMap(), IsMapWithSize.aMapWithSize(1));
+        assertThat(service.getCryptoKeyProviderMap(), IsMapContaining.hasEntry(new CaseInsensitiveString("file"),
+                mockKeyProvider));
+    }
+
+    @Test
+    void GIVEN_key_service_providers_with_same_key_type_WHEN_register_provider_THEN_throw_exception() throws Exception {
+        when(mockKeyProvider.supportedKeyType()).thenReturn("FILE");
+        service.registerCryptoKeyProvider(mockKeyProvider);
+        CryptoKeySpi providerB = mock(CryptoKeySpi.class);
+        when(providerB.supportedKeyType()).thenReturn("file");
+        assertThrows(ServiceProviderConflictException.class, () -> service.registerCryptoKeyProvider(providerB));
+    }
+
+    @Test
+    void GIVEN_key_service_provider_registered_WHEN_deregister_THEN_removed() throws Exception {
+        when(mockKeyProvider.supportedKeyType()).thenReturn("FILE");
+        service.registerCryptoKeyProvider(mockKeyProvider);
+        service.deregisterCryptoKeyProvider(mockKeyProvider);
+        assertThat(service.getCryptoKeyProviderMap(), IsMapWithSize.aMapWithSize(0));
```

Comment:
de register it twice to validate that case as well?

Reply:
will do